File: src/prediction/vocabulary_pack.py
"""
Vocabulary Pack system for domain-specific prediction.

Packs are additive layers of vocabulary (unigrams, bigrams, trigrams)
that users can enable/disable independently of their accessibility profile.

Pack format (on disk):
    data/packs/<pack-name>/
    ├── pack.json          # Metadata (name, description, version)
    ├── dictionary.txt     # One word per line
    ├── bigrams.txt        # word1 word2 per line (optional)
    └── trigrams.txt       # word1 word2 word3 per line (optional)
"""

from __future__ import annotations

import json
import logging
import shutil
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Optional, Set

_logger = logging.getLogger("VocabularyPack")

# Weight for pack vocabulary (lower than user-learned so personal typing wins)
PACK_UNIGRAM_WEIGHT = 3
PACK_BIGRAM_WEIGHT = 30
PACK_TRIGRAM_WEIGHT = 30


@dataclass
class VocabularyPack:
    """A loadable vocabulary pack for a specific domain."""

    name: str
    description: str
    path: Path
    version: int = 1
    enabled: bool = False

    # Loaded data
    words: Set[str] = field(default_factory=set)
    bigrams: Dict[str, Dict[str, int]] = field(default_factory=dict)
    trigrams: Dict[str, Dict[str, int]] = field(default_factory=dict)

    @classmethod
    def from_directory(cls, pack_dir: Path) -> Optional[VocabularyPack]:
        """
        Load a vocabulary pack from a directory.

        Args:
            pack_dir: Path to the pack directory

        Returns:
            VocabularyPack if valid, None if missing/invalid
        """
        if not pack_dir.is_dir():
            return None

        # Load metadata
        meta_path = pack_dir / "pack.json"
        if meta_path.exists():
            try:
                meta = json.loads(meta_path.read_text())
                name = meta.get("name", pack_dir.name)
                description = meta.get("description", "")
                version = meta.get("version", 1)
            except (json.JSONDecodeError, OSError):
                name = pack_dir.name
                description = ""
                version = 1
        else:
            name = pack_dir.name
            description = ""
            version = 1

        pack = cls(
            name=name,
            description=description,
            path=pack_dir,
            version=version,
        )
        return pack

    def load(self) -> bool:
        """
        Load vocabulary data from disk into memory.

        Returns:
            True if any data was loaded
        """
        loaded_any = False

        # Load dictionary (one word per line)
        dict_path = self.path / "dictionary.txt"
        if dict_path.exists():
            try:
                with open(dict_path) as f:
                    for line in f:
                        word = line.strip().lower()
                        if word and not word.startswith("#"):
                            self.words.add(word)
                loaded_any = bool(self.words)
                _logger.info("Pack '%s': loaded %d words", self.name, len(self.words))
            except OSError as e:
                _logger.error("Pack '%s': failed to load dictionary: %s", self.name, e)

        # Load bigrams
        bigrams_path = self.path / "bigrams.txt"
        if bigrams_path.exists():
            try:
                count = 0
                with open(bigrams_path) as f:
                    for line in f:
                        line = line.strip()
                        if not line or line.startswith("#"):
                            continue
                        parts = line.split()
                        if len(parts) >= 2:
                            w1, w2 = parts[0].lower(), parts[1].lower()
                            if w1 not in self.bigrams:
                                self.bigrams[w1] = {}
                            self.bigrams[w1][w2] = self.bigrams[w1].get(w2, 0) + PACK_BIGRAM_WEIGHT
                            count += 1
                _logger.info("Pack '%s': loaded %d bigrams", self.name, count)
                loaded_any = loaded_any or count > 0
            except OSError as e:
                _logger.error("Pack '%s': failed to load bigrams: %s", self.name, e)

        # Load trigrams
        trigrams_path = self.path / "trigrams.txt"
        if trigrams_path.exists():
            try:
                count = 0
                with open(trigrams_path) as f:
                    for line in f:
                        line = line.strip()
                        if not line or line.startswith("#"):
                            continue
                        parts = line.split()
                        if len(parts) >= 3:
                            w1, w2, w3 = parts[0].lower(), parts[1].lower(), parts[2].lower()
                            key = f"{w1} {w2}"
                            if key not in self.trigrams:
                                self.trigrams[key] = {}
                            self.trigrams[key][w3] = (
                                self.trigrams[key].get(w3, 0) + PACK_TRIGRAM_WEIGHT
                            )
                            count += 1
                _logger.info("Pack '%s': loaded %d trigrams", self.name, count)
                loaded_any = loaded_any or count > 0
            except OSError as e:
                _logger.error("Pack '%s': failed to load trigrams: %s", self.name, e)

        return loaded_any

    def unload(self) -> None:
        """Clear loaded data from memory."""
        self.words.clear()
        self.bigrams.clear()
        self.trigrams.clear()

    def get_info(self) -> dict:
        """Return pack metadata for UI display."""
        return {
            "name": self.name,
            "description": self.description,
            "version": self.version,
            "enabled": self.enabled,
            "words": len(self.words),
            "bigrams": sum(len(v) for v in self.bigrams.values()),
            "trigrams": sum(len(v) for v in self.trigrams.values()),
        }


class PackManager:
    """
    Manages vocabulary packs — discovery, loading, and application to the predictor.

    Packs are discovered from ``data/packs/`` on init. Users can enable/disable
    packs at runtime. Enabled packs inject their vocabulary into the predictor's
    n-gram model.
    """

    def __init__(self, packs_dir: Optional[Path] = None,
                 user_packs_dir: Optional[Path] = None):
        """
        Initialize the pack manager.

        Args:
            packs_dir: Directory containing built-in pack subdirectories.
                       Defaults to ``data/packs/`` relative to project root.
            user_packs_dir: Directory for user-imported custom packs.
                           Defaults to ``<config_dir>/packs/``.
        """
        if packs_dir is None:
            packs_dir = Path(__file__).parent.parent.parent / "data" / "packs"

        if user_packs_dir is None:
            from ..platform import get_config_dir
            user_packs_dir = get_config_dir() / "packs"
            user_packs_dir.mkdir(parents=True, exist_ok=True)

        self._packs_dir = packs_dir
        self._user_packs_dir = user_packs_dir
        self._packs: Dict[str, VocabularyPack] = {}

        self._discover_packs()

    def _discover_packs(self) -> None:
        """Scan both built-in and user packs directories."""
        for packs_dir in (self._packs_dir, self._user_packs_dir):
            if not packs_dir.is_dir():
                continue
            for item in sorted(packs_dir.iterdir()):
                if item.is_dir() and not item.name.startswith("."):
                    pack = VocabularyPack.from_directory(item)
                    if pack is not None and item.name not in self._packs:
                        self._packs[item.name] = pack
                        _logger.info("Discovered pack: %s (%s)", pack.name, packs_dir)

        _logger.info("Found %d vocabulary packs", len(self._packs))

    def get_available_packs(self) -> List[str]:
        """Return list of available pack IDs."""
        return list(self._packs.keys())

    def get_pack_info(self, pack_id: str) -> Optional[dict]:
        """Return metadata for a pack."""
        pack = self._packs.get(pack_id)
        if pack:
            return pack.get_info()
        return None

    def get_all_pack_info(self) -> List[dict]:
        """Return metadata for all packs."""
        return [pack.get_info() for pack in self._packs.values()]

    def get_enabled_packs(self) -> List[str]:
        """Return list of enabled pack IDs."""
        return [pid for pid, pack in self._packs.items() if pack.enabled]

    def enable_pack(self, pack_id: str) -> bool:
        """
        Enable a pack and load its data.

        Args:
            pack_id: Pack directory name

        Returns:
            True if pack was found and enabled
        """
        pack = self._packs.get(pack_id)
        if pack is None:
            _logger.warning("Pack not found: %s", pack_id)
            return False

        if pack.enabled:
            return True  # Already enabled

        pack.load()
        pack.enabled = True
        _logger.info("Enabled pack: %s (%d words)", pack.name, len(pack.words))
        return True

    def disable_pack(self, pack_id: str) -> bool:
        """
        Disable a pack and unload its data.

        Args:
            pack_id: Pack directory name

        Returns:
            True if pack was found and disabled
        """
        pack = self._packs.get(pack_id)
        if pack is None:
            return False

        if not pack.enabled:
            return True  # Already disabled

        pack.unload()
        pack.enabled = False
        _logger.info("Disabled pack: %s", pack.name)
        return True

    def apply_to_predictor(self, predictor) -> None:
        """
        Inject all enabled packs' vocabulary into a predictor.

        Called by HybridPredictor when predictions are generated.
        Only applies packs that are currently enabled and loaded.

        Args:
            predictor: NgramPredictor instance to inject into
        """
        for pack in self._packs.values():
            if not pack.enabled:
                continue

            # Inject unigrams
            for word in pack.words:
                predictor.unigrams[word] = max(
                    predictor.unigrams.get(word, 0), PACK_UNIGRAM_WEIGHT
                )

            # Inject bigrams
            for w1, targets in pack.bigrams.items():
                for w2, weight in targets.items():
                    predictor.bigrams[w1][w2] = max(
                        predictor.bigrams[w1].get(w2, 0), weight
                    )

            # Inject trigrams
            for key, targets in pack.trigrams.items():
                for w3, weight in targets.items():
                    predictor.trigrams[key][w3] = max(
                        predictor.trigrams[key].get(w3, 0), weight
                    )

    def import_pack(self, source_dir: Path) -> Optional[str]:
        """
        Import a vocabulary pack from an external directory.

        Copies the pack folder into the user packs directory. The source
        must contain at least a ``dictionary.txt`` file. A ``pack.json``
        is optional (the folder name is used as the pack ID).

        Args:
            source_dir: Path to the pack folder to import.

        Returns:
            The pack ID (folder name) on success, None on failure.
        """
        source_dir = Path(source_dir)
        if not source_dir.is_dir():
            _logger.error("Import source is not a directory: %s", source_dir)
            return None

        # Must have at least a dictionary.txt
        if not (source_dir / "dictionary.txt").exists():
            _logger.error("Import source missing dictionary.txt: %s", source_dir)
            return None

        pack_id = source_dir.name.lower().replace(" ", "_")
        dest_dir = self._user_packs_dir / pack_id

        # Don't overwrite built-in packs
        if pack_id in self._packs and (self._packs_dir / pack_id).is_dir():
            _logger.error("Cannot overwrite built-in pack: %s", pack_id)
            return None

        try:
            if dest_dir.exists():
                shutil.rmtree(dest_dir)
            shutil.copytree(source_dir, dest_dir)

            # Generate pack.json if missing
            if not (dest_dir / "pack.json").exists():
                meta = {
                    "name": source_dir.name,
                    "description": f"Custom pack imported from {source_dir.name}",
                    "version": 1,
                }
                (dest_dir / "pack.json").write_text(json.dumps(meta))

            # Register the new pack
            pack = VocabularyPack.from_directory(dest_dir)
            if pack is not None:
                self._packs[pack_id] = pack
                _logger.info("Imported pack: %s (%d)", pack.name, len(list(dest_dir.iterdir())))
                return pack_id

        except Exception as e:
            _logger.error("Failed to import pack from %s: %s", source_dir, e)

        return None

    def get_user_packs_dir(self) -> Path:
        """Return the user packs directory path."""
        return self._user_packs_dir

    def get_pack(self, pack_id: str) -> Optional[VocabularyPack]:
        """Get a pack by ID (for testing/direct access)."""
        return self._packs.get(pack_id)

File: cpp/prediction/VocabularyPack.cpp
#include "VocabularyPack.h"

#include "NgramPredictor.h"
#include "../Paths.h"

#include <QDebug>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonDocument>
#include <QJsonObject>
#include <QRegularExpression>
#include <QTextStream>

namespace {

QStringList readPackLines(const QString &path)
{
    QStringList out;
    QFile f(path);
    if (!f.open(QIODevice::ReadOnly | QIODevice::Text))
        return out;
    QTextStream ts(&f);
    while (!ts.atEnd())
        out << ts.readLine();
    return out;
}

} // namespace

// ===== VocabularyPack ====================================================

std::unique_ptr<VocabularyPack> VocabularyPack::fromDirectory(const QString &packDir)
{
    QFileInfo fi(packDir);
    if (!fi.isDir())
        return nullptr;

    auto pack = std::make_unique<VocabularyPack>();
    pack->m_path = fi.absoluteFilePath();
    pack->m_name = fi.fileName();
    pack->m_description.clear();
    pack->m_version = 1;

    const QString metaPath = QDir(packDir).filePath("pack.json");
    if (QFileInfo::exists(metaPath)) {
        QFile f(metaPath);
        if (f.open(QIODevice::ReadOnly)) {
            QJsonParseError err{};
            const QJsonDocument doc = QJsonDocument::fromJson(f.readAll(), &err);
            if (err.error == QJsonParseError::NoError && doc.isObject()) {
                const QJsonObject o = doc.object();
                pack->m_name = o.value("name").toString(pack->m_name);
                pack->m_description = o.value("description").toString();
                pack->m_version = o.value("version").toInt(1);
            }
        }
    }
    return pack;
}

QString VocabularyPack::id() const
{
    return QFileInfo(m_path).fileName();
}

bool VocabularyPack::load()
{
    bool loadedAny = false;

    for (const QString &raw : readPackLines(QDir(m_path).filePath("dictionary.txt"))) {
        const QString word = raw.trimmed().toLower();
        if (!word.isEmpty() && !word.startsWith('#'))
            m_words.insert(word);
    }
    loadedAny = !m_words.isEmpty();

    for (const QString &raw : readPackLines(QDir(m_path).filePath("bigrams.txt"))) {
        const QString line = raw.trimmed();
        if (line.isEmpty() || line.startsWith('#'))
            continue;
        const QStringList parts = line.split(QRegularExpression("\\s+"), Qt::SkipEmptyParts);
        if (parts.size() >= 2) {
            const QString w1 = parts[0].toLower(), w2 = parts[1].toLower();
            m_bigrams[w1][w2] = m_bigrams[w1].value(w2, 0) + kBigramWeight;
            loadedAny = true;
        }
    }

    for (const QString &raw : readPackLines(QDir(m_path).filePath("trigrams.txt"))) {
        const QString line = raw.trimmed();
        if (line.isEmpty() || line.startsWith('#'))
            continue;
        const QStringList parts = line.split(QRegularExpression("\\s+"), Qt::SkipEmptyParts);
        if (parts.size() >= 3) {
            const QString key = parts[0].toLower() + " " + parts[1].toLower();
            const QString w3 = parts[2].toLower();
            m_trigrams[key][w3] = m_trigrams[key].value(w3, 0) + kTrigramWeight;
            loadedAny = true;
        }
    }
    return loadedAny;
}

void VocabularyPack::unload()
{
    m_words.clear();
    m_bigrams.clear();
    m_trigrams.clear();
}

QVariantMap VocabularyPack::getInfo() const
{
    int bi = 0;
    for (const auto &t : m_bigrams)
        bi += t.size();
    int tri = 0;
    for (const auto &t : m_trigrams)
        tri += t.size();
    QVariantMap m;
    m.insert("id", id());
    m.insert("name", m_name);
    m.insert("description", m_description);
    m.insert("version", m_version);
    m.insert("enabled", m_enabled);
    m.insert("words", m_words.size());
    m.insert("bigrams", bi);
    m.insert("trigrams", tri);
    return m;
}

// ===== PackManager =======================================================

PackManager::PackManager(const QString &builtinPacksDir, const QString &userPacksDir)
    : m_builtinPacksDir(builtinPacksDir.isEmpty() ? QDir(paths::dataDir()).filePath("packs")
                                                  : builtinPacksDir)
    , m_userPacksDir(userPacksDir.isEmpty() ? QDir(paths::configDir()).filePath("packs")
                                            : userPacksDir)
{
    QDir().mkpath(m_userPacksDir);
    discoverPacks();
}

void PackManager::discoverPacks()
{
    for (const QString &dir : {m_builtinPacksDir, m_userPacksDir}) {
        QDir d(dir);
        if (!d.exists())
            continue;
        const QStringList subdirs = d.entryList(QDir::Dirs | QDir::NoDotAndDotDot, QDir::Name);
        for (const QString &name : subdirs) {
            if (name.startsWith('.') || m_packs.contains(name))
                continue;
            auto pack = VocabularyPack::fromDirectory(d.filePath(name));
            if (pack) {
                m_packs.insert(name, std::shared_ptr<VocabularyPack>(std::move(pack)));
                m_order << name;
            }
        }
    }
}

QStringList PackManager::getAvailablePacks() const
{
    return m_order;
}

QVariantList PackManager::getAllPackInfo() const
{
    QVariantList out;
    for (const QString &id : m_order)
        out.append(m_packs.value(id)->getInfo());
    return out;
}

QStringList PackManager::getEnabledPacks() const
{
    QStringList out;
    for (const QString &id : m_order)
        if (m_packs.value(id)->enabled())
            out << id;
    return out;
}

bool PackManager::enablePack(const QString &id)
{
    auto it = m_packs.constFind(id);
    if (it == m_packs.constEnd())
        return false;
    if (it.value()->enabled())
        return true;
    it.value()->load();
    it.value()->setEnabled(true);
    return true;
}

bool PackManager::disablePack(const QString &id)
{
    auto it = m_packs.constFind(id);
    if (it == m_packs.constEnd())
        return false;
    if (!it.value()->enabled())
        return true;
    it.value()->unload();
    it.value()->setEnabled(false);
    return true;
}

void PackManager::applyToPredictor(NgramPredictor *predictor) const
{
    if (!predictor)
        return;
    for (const QString &id : m_order) {
        const auto &pack = m_packs.value(id);
        if (pack && pack->enabled())
            predictor->injectVocab(pack->words(), VocabularyPack::kUnigramWeight,
                                   pack->bigrams(), pack->trigrams());
    }
}

bool PackManager::copyTreeSkipSymlinks(const QString &src, const QString &dst)
{
    QDir().mkpath(dst);
    QDir srcDir(src);
    const QFileInfoList entries =
        srcDir.entryInfoList(QDir::AllEntries | QDir::NoDotAndDotDot | QDir::Hidden | QDir::System);
    for (const QFileInfo &fi : entries) {
        if (fi.isSymLink())
            continue; // never dereference symlinks out of the pack tree
        const QString destPath = QDir(dst).filePath(fi.fileName());
        if (fi.isDir()) {
            if (!copyTreeSkipSymlinks(fi.absoluteFilePath(), destPath))
                return false;
        } else if (fi.isFile()) {
            if (!QFile::copy(fi.absoluteFilePath(), destPath))
                return false;
        }
    }
    return true;
}

QString PackManager::importPack(const QString &sourceDir)
{
    QFileInfo si(sourceDir);
    if (!si.isDir()) {
        qWarning() << "Import source is not a directory:" << sourceDir;
        return QString();
    }
    if (!QFileInfo::exists(QDir(sourceDir).filePath("dictionary.txt"))) {
        qWarning() << "Import source missing dictionary.txt:" << sourceDir;
        return QString();
    }

    // Derive + sanitise the pack id (filesystem directory name).
    QString rawId = si.fileName().toLower().replace(' ', '_');
    QString packId;
    for (QChar c : rawId) {
        const char l = c.toLatin1();
        if ((l >= 'a' && l <= 'z') || (l >= '0' && l <= '9') || l == '_' || l == '-')
            packId += c;
    }
    while (!packId.isEmpty() && (packId.front() == '_' || packId.front() == '-'))
        packId.remove(0, 1);
    while (!packId.isEmpty() && (packId.back() == '_' || packId.back() == '-'))
        packId.chop(1);

    static const QRegularExpression valid("^[a-z0-9][a-z0-9_\\-]{0,63}$");
    if (packId.isEmpty() || !valid.match(packId).hasMatch()) {
        qWarning() << "Rejected pack import: invalid id derived from" << sourceDir;
        return QString();
    }

    // Defence in depth: the resolved destination must sit strictly under the
    // user packs dir.
    const QString rootAbs = QDir(m_userPacksDir).absolutePath();
    const QString destAbs = QDir::cleanPath(QDir(m_userPacksDir).absoluteFilePath(packId));
    if (QDir(rootAbs).relativeFilePath(destAbs) != packId) {
        qWarning() << "Rejected pack import: destination escapes" << rootAbs;
        return QString();
    }

    // Don't overwrite a built-in pack.
    if (m_packs.contains(packId) && QFileInfo(QDir(m_builtinPacksDir).filePath(packId)).isDir()) {
        qWarning() << "Cannot overwrite built-in pack:" << packId;
        return QString();
    }

    QDir destDir(destAbs);
    if (destDir.exists())
        destDir.removeRecursively();
    if (!copyTreeSkipSymlinks(si.absoluteFilePath(), destAbs)) {
        qWarning() << "Failed to copy pack into" << destAbs;
        return QString();
    }

    if (!QFileInfo::exists(QDir(destAbs).filePath("pack.json"))) {
        QJsonObject meta;
        meta.insert("name", si.fileName());
        meta.insert("description", QStringLiteral("Custom pack imported from %1").arg(si.fileName()));
        meta.insert("version", 1);
        QFile f(QDir(destAbs).filePath("pack.json"));
        if (f.open(QIODevice::WriteOnly))
            f.write(QJsonDocument(meta).toJson(QJsonDocument::Compact));
    }

    auto pack = VocabularyPack::fromDirectory(destAbs);
    if (!pack)
        return QString();
    m_packs.insert(packId, std::shared_ptr<VocabularyPack>(std::move(pack)));
    if (!m_order.contains(packId))
        m_order << packId;
    return packId;
}
